mapper/pchain/tx_parser.go

```diff
@@ -0,0 +1,883 @@
+	case *txs.ImportTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.AddValidatorTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.AddPermissionlessValidatorTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.AddDelegatorTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.AddPermissionlessDelegatorTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.CreateSubnetTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.CreateChainTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.AddSubnetValidatorTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.TransformSubnetTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.RemoveSubnetValidatorTx:
+		txIds = append(txIds, getUniqueTxIds(unsignedTx.Ins)...)
+	case *txs.RewardValidatorTx:
+		txIds = append(txIds, unsignedTx.TxID)
+	case *txs.AdvanceTimeTx:
+		// advance time txs do not have inputs
+	default:
+		log.Printf("unknown type %T", unsignedTx)
+	}
+
+	ids.SortIDs(txIds)
+
+	return txIds, nil
+}
+
+func getUniqueTxIds(ins []*avax.TransferableInput) []ids.ID {
+	txnIDs := make(map[string]ids.ID)
+	for _, in := range ins {
+		txnIDs[in.UTXOID.TxID.String()] = in.UTXOID.TxID
+	}
+
+	uniqueTxnIDs := make([]ids.ID, 0, len(txnIDs))
+	for _, txnID := range txnIDs {
+		uniqueTxnIDs = append(uniqueTxnIDs, txnID)
+	}
+	return uniqueTxnIDs
+}
+
+func getUTXOMap(d *DependencyTx) map[uint32]*avax.UTXO {
+	utxos := make(map[uint32]*avax.UTXO)
+
+	if d.Tx != nil {
+		// Generate UTXOs from outputs
+		switch unsignedTx := d.Tx.Unsigned.(type) {
+		case *txs.ExportTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		case *txs.ImportTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		case *txs.AddValidatorTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+			mapUTXOs(d.Tx.ID(), unsignedTx.Stake(), utxos)
+		case *txs.AddPermissionlessValidatorTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+			mapUTXOs(d.Tx.ID(), unsignedTx.Stake(), utxos)
+		case *txs.AddDelegatorTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+			mapUTXOs(d.Tx.ID(), unsignedTx.Stake(), utxos)
+		case *txs.AddPermissionlessDelegatorTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+			mapUTXOs(d.Tx.ID(), unsignedTx.Stake(), utxos)
+		case *txs.CreateSubnetTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		case *txs.AddSubnetValidatorTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		case *txs.TransformSubnetTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		case *txs.RemoveSubnetValidatorTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		case *txs.CreateChainTx:
+			mapUTXOs(d.Tx.ID(), unsignedTx.Outs, utxos)
+		default:
+			// no utxos extracted from unsupported transaction types
+		}
+	}
+
+	// Add reward UTXOs
+	for _, utxo := range d.RewardUTXOs {
+		utxos[utxo.OutputIndex] = utxo
```

Comment:
@bilgehansahin I doubt this is correct; I think this op may overwrite some of the utxos previously filled.
Call `A` the tx whose dependency are represented by `d` and call `B` `d.Tx` i.e. the tx carrying the utxos spent by `A`
We fill `utxos` with two data:

1. all utxos from `B`
2. RewardUTXO which is built by avalanchego as a utxo from `A` with index len(A.utxos)+len(A.stakes).

So if B has many more utxos than A, this op would overwrite the a B utxo.
I believe we need to handle these dependencies totally differently

Reply:
Sure 👍